sdks/java/core/src/test/java/org/apache/beam/sdk/fn/data/BeamFnDataInboundObserverTest.java

```diff
@@ -165,6 +165,7 @@ public void testCloseVisibleToAwaitCompletionCallerAndProducer() throws Exceptio
     Future<?> future2 =
         executor.submit(
             () -> {
+              Thread.sleep(500); // give the previous future some time to reach the inf while loop
```

Comment:
Busy sleep is generally not preferable. Consider use wait/notify which sounds exactly what we need here? That is block this thread until future1 reaches while(true)